### PR TITLE
Tab complete for chat-cli

### PR DESCRIPTION
**Pull request by liam-fitzgerald**
Adds tab complete for chat-cli with a static list of supported commands

### CODE REVIEW

File: pkg/arvo/app/chat-cli.hoon [hunk 1]
```diff
@@ -309,8 +309,59 @@
...
+      [%set leaf+";set key value"]
+      [%unset leaf+";unset key"]
+    ::
+      [%chats leaf+";chats"]
+      [%help leaf+";help"]
+    ==
+  ++  tab
+    |=  pos=@ud
+    ^-  (quip move _this)
+    ?.  =(';' (snag 0 buf.state.cli))
```
[the remark this answers is not in the document]
Comment (liam-fitzgerald) [reply pkg/arvo/app/chat-cli.hoon]: get-id won't include the `;` so check it to make sure that we don't try and autocomplete on prose

File: pkg/arvo/app/chat-cli.hoon [hunk 2]
```diff
@@ -309,8 +309,59 @@
...
+      %tab  (tab +.act)
     ==
+  ::  +tab-list: static list of autocomplete entries
+  ++  tab-list
+    ^-  (list (option:auto tank))
+    :~
+      [%join leaf+";join ~ship/chat-name"]
+      [%leave leaf+";leave ~ship/chat-name"]
+      ::
+      [%create leaf+";create [type] /chat-name [glyph]"]
```
[the remark this answers is not in the document]
Comment (liam-fitzgerald) [reply pkg/arvo/app/chat-cli.hoon]: how do you propose indicating it's optionality?

[the remark this answers is not in the document]
Comment (Fang-) [reply pkg/arvo/app/chat-cli.hoon]: Probably `(glyph)`? And then doing `[r|w|rw]` for the required multiple-choice for `;invite` and `;banish`.

File: pkg/arvo/app/chat-cli.hoon [hunk 3]
```diff
@@ -309,8 +309,59 @@
...
+      [(tab:sh-out options) moves]
+    =|  fxs=(list sole-effect:sole-sur)
+    |-
+    ?~  to-send
+      [(flop moves) this]
+    =^  char  state.cli
+      (~(transmit sole-lib state.cli) [%ins send-pos `@c`i.to-send])
+    $(moves [(effect:sh-out %det char) moves], send-pos +(send-pos), to-send t.to-send)
+
+
```
Comment (loganallenc) [inline pkg/arvo/app/chat-cli.hoon]: empty lines?

File: pkg/arvo/app/chat-cli.hoon [hunk 4]
```diff
@@ -931,6 +982,13 @@
     |=  fec=sole-effect:sole-sur
     ^-  move
     [bone.cli %diff %sole-effect fec]
+  ::  +tab: print tab-complete list
+  ::
+  ++  tab
+    |=  options=(list [cord tank])
+    ^-  move
+    (effect %tab options)
+
```
Comment (loganallenc) [inline pkg/arvo/app/chat-cli.hoon]: empty?

**Comment by philipcmonk** [comment]
This isn't going to make it in for the breach, but it should be an easy OTA afterward.  @Fang- or @liam-fitzgerald, can one of you migrate these changes to the `rc` branch?

**Comment by liam-fitzgerald** [comment]
Happy to do it

**Comment by philipcmonk** [comment]
The update for static gall looks sane to me.

**Comment by philipcmonk** [comment]
Doesn't need to go in this PR, but someone mentioned on urbit-help that it'd be cool to tab-complete in a message on `~` by reverse-searching incoming messages for a name.  So if ~wicdev-wisryt just sent a message, ~w<TAB> would expand to that.

**Pull request merged**